File: frontend/src/stores/update.ts
import { defineStore } from 'pinia'
import {
  api,
  type UpdateCheckResult,
  type UpdateDownloadEvent,
  type PendingUpdateInfo,
} from '../api/bindings'

// useUpdateStore 管理更新检查与下载进度。
// 后端不通过 Wails 事件推送状态（见 AGENTS.md），前端轮询绑定方法获取进度。
export const useUpdateStore = defineStore('update', {
  state: () => ({
    version: '',
    result: null as UpdateCheckResult | null,
    progress: null as UpdateDownloadEvent | null,
    pending: null as PendingUpdateInfo | null,
    checking: false,
    downloading: false,
  }),
  getters: {
    hasUpdate: (s) => s.result?.status === 'update-available',
    latestVersion: (s) => s.result?.latestVersion ?? '',
    pendingVersion: (s) => (s.pending?.exists ? s.pending.version : ''),
    downloadPercent: (s) => s.progress?.percent ?? 0,
    canRestartUpdate: (s) => !!s.pending?.exists || s.progress?.phase === 'cancelled',
  },
  actions: {
    async loadVersion() {
      this.version = await api.getVersion()
    },
    async loadPending() {
      this.pending = await api.pendingUpdateInfo()
    },
    async check() {
      this.checking = true
      try {
        this.result = await api.checkUpdate()
        return this.result
      } finally {
        this.checking = false
      }
    },
    async download() {
      const err = await api.downloadAndApplyUpdate()
      if (err) throw new Error(err)
      this.downloading = true
      this.pollProgress()
    },
    pollProgress() {
      this.progress = null
      const timer = setInterval(async () => {
        const ev = await api.updateProgress()
        this.progress = ev
        if (ev.phase === 'completed' || ev.phase === 'error' || ev.phase === 'cancelled') {
          clearInterval(timer)
          this.downloading = false
          await this.loadPending()
        }
      }, 500)
    },
    async applyRestart(): Promise<string> {
      const msg = await api.applyUpdateAndRestart()
      if (!msg) await this.loadPending()
      return msg
    },
  },
})

File: model/update.go
package model

// 更新检查状态常量。
const (
	UpdateStatusUpToDate  = "up-to-date"
	UpdateStatusAvailable = "update-available"
	UpdateStatusError     = "error"
)

// UpdateCheckResult 是一次更新检查的结果。
type UpdateCheckResult struct {
	// Status 取值：up-to-date / update-available / error。
	Status         string `json:"status"`
	CurrentVersion string `json:"currentVersion"`
	LatestVersion  string `json:"latestVersion"`
	ReleaseURL     string `json:"releaseUrl"`
	// DownloadURL 是 workbuddy-checkin.exe 资产的下载地址；资产缺失时为空。
	DownloadURL string `json:"downloadUrl"`
	Message     string `json:"message"`
}

// 更新下载事件阶段常量。
const (
	UpdateDownloadPhaseDownloading = "downloading"
	UpdateDownloadPhaseCompleted   = "completed"
	UpdateDownloadPhaseCancelled   = "cancelled"
	UpdateDownloadPhaseError       = "error"
)

// UpdateDownloadEvent 是一次更新下载的进度事件，通过 Wails events（update:progress）推送给前端。
type UpdateDownloadEvent struct {
	// Phase 取值：downloading / completed / cancelled / error。
	Phase      string `json:"phase"`
	Downloaded int64  `json:"downloaded"`
	Total      int64  `json:"total"`
	Percent    int    `json:"percent"`
	// Message 为错误阶段的提示文案。
	Message string `json:"message"`
}

// PendingUpdateInfo 描述已下载待应用（.new 就绪）的更新状态。
type PendingUpdateInfo struct {
	// Exists 表示 exe 同目录是否存在 .new 待应用文件。
	Exists bool `json:"exists"`
	// Version 是 .new 对应的版本号（来自 .new.version 元数据），读取失败时为空。
	Version string `json:"version"`
}
